File: src/programmers/BS_Immigration.py
"""
https://programmers.co.kr/learn/courses/30/lessons/43238/
Using Binary Search
Search minimal value that meets the condition.
"""
def solution(n, times):
    times.sort()
    left, right = 0, int(1E20)

    while left <= right:
        mid = (left + right) // 2

        if binary_search(times, mid, n):
            right = mid - 1
        else:
            left = mid + 1

    return left

def binary_search(times, ans, n):
    count = 0
    # Check if the immigration can ends with given time
    for t in times:
        count += ans // t  # Calculate people numbers by one officer

        if count >= n:
            return True

    return False

File: src/programmers/BS_Immigration.cpp
"""
https://programmers.co.kr/learn/courses/30/lessons/43238?language=cpp
Using binary search
"""
#1. My Solution
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool binary_search(vector <int> times, long long ans, long long n) {
    long long count = 0;
    for (auto t: times) {
        count += ans / t;
    }
    
    if (count >= n) {
        return true;
    }
    
    return false;
}

long long solution(int n, vector<int> times) {
    sort(times.begin(), times.end());
    long long left, right, mid;
    left = 0;
    right = 1E15;
    
    while (left <= right) {
        mid = (left + right) / 2;
        
        if (binary_search(times, mid, n)) {
            right = mid - 1;
        }
        else {
            left = mid + 1;
        }   
    }
    
    return left;
}
